File: zero-server/src/main/kotlin/com/zeroday/handler/handlers/AuctionHandler.kt
package com.zeroday.handler.handlers

import com.zeroday.handler.HandlerContext
import com.zeroday.handler.HandlerResult
import com.zeroday.handler.MessageHandler
import com.zeroday.handler.str
import com.zeroday.protocol.ErrorCategory
import com.zeroday.protocol.RequestTypes
import com.zeroday.protocol.ResponseTypes
import com.zeroday.protocol.ServerError
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonPrimitive

class AuctionHandler : MessageHandler {
    override val handledTypes: Set<String> = setOf(
        RequestTypes.AUCTION_LIST, RequestTypes.AUCTION_SEARCH, RequestTypes.AUCTION_CREATE,
        RequestTypes.AUCTION_BID, RequestTypes.AUCTION_BUYOUT,
        RequestTypes.AUCTION_MY_LISTINGS, RequestTypes.AUCTION_CANCEL
    )

    override val requiresAuth: Boolean = true

    override suspend fun handle(ctx: HandlerContext, type: String, payload: JsonObject): HandlerResult {
        val playerId = ctx.authenticatedPlayerId()
            ?: return HandlerResult.Err(ServerError.fromMessage("Authentication required", ErrorCategory.AUTHENTICATION))
        return when (type) {
            RequestTypes.AUCTION_LIST -> listActive(ctx)
            RequestTypes.AUCTION_SEARCH -> search(ctx, payload)
            RequestTypes.AUCTION_CREATE -> create(ctx, playerId, payload)
            RequestTypes.AUCTION_BID -> bid(ctx, playerId, payload)
            RequestTypes.AUCTION_BUYOUT -> buyout(ctx, playerId, payload)
            RequestTypes.AUCTION_MY_LISTINGS -> myListings(ctx, playerId)
            RequestTypes.AUCTION_CANCEL -> cancel(ctx, playerId, payload)
            else -> HandlerResult.Err(ServerError.fromMessage("Unknown auction request", ErrorCategory.VALIDATION))
        }
    }

    private fun JsonObject.strOr(key: String, default: String = ""): String =
        (this[key] as? JsonPrimitive)?.content ?: default

    private fun JsonObject.longOr(key: String, default: Long = 0L): Long =
        (this[key] as? JsonPrimitive)?.content?.toLongOrNull() ?: default

    private fun JsonObject.intOr(key: String, default: Int = 0): Int =
        (this[key] as? JsonPrimitive)?.content?.toIntOrNull() ?: default

    private suspend fun listActive(ctx: HandlerContext): HandlerResult {
        val listings = ctx.services.auctionService.searchListings()
        return HandlerResult.Ok(ResponseTypes.AUCTION_LISTINGS, mapOf(
            "listings" to listings.map { ctx.services.auctionService.serializeListing(it) },
            "count" to listings.size
        ))
    }

    private suspend fun search(ctx: HandlerContext, payload: JsonObject): HandlerResult {
        val results = ctx.services.auctionService.searchListings(
            query = payload.strOr("query", "").ifBlank { null },
            rarity = payload.strOr("rarity", "").ifBlank { null },
            slot = payload.strOr("slot", "").ifBlank { null },
            minPrice = payload.longOr("minPrice").takeIf { it > 0 },
            maxPrice = payload.longOr("maxPrice").takeIf { it > 0 }
        )
        return HandlerResult.Ok(ResponseTypes.AUCTION_SEARCH_RESULTS, mapOf(
            "results" to results.map { ctx.services.auctionService.serializeListing(it) },
            "count" to results.size
        ))
    }

    private suspend fun create(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val itemName = payload.str("itemName")
            ?: return HandlerResult.Err(ServerError.fromMessage("itemName required", ErrorCategory.VALIDATION))
        val itemId = payload.str("itemId")
            ?: return HandlerResult.Err(ServerError.fromMessage("itemId required", ErrorCategory.VALIDATION))
        val startingBid = payload.longOr("startingBid")
        if (startingBid <= 0) return HandlerResult.Err(ServerError.fromMessage("startingBid required", ErrorCategory.VALIDATION))

        val result = ctx.services.auctionService.createListing(
            sellerId = playerId,
            itemName = itemName,
            itemId = itemId,
            itemRarity = payload.strOr("itemRarity", "Common"),
            itemSlot = payload.strOr("itemSlot", "Any"),
            itemTier = payload.intOr("itemTier", 1),
            itemLevel = payload.intOr("itemLevel", 1),
            startingBid = startingBid,
            buyoutPrice = payload.longOr("buyoutPrice"),
            durationHours = payload.intOr("durationHours", 24)
        )
        return result.fold(
            onSuccess = { listing -> HandlerResult.Ok(ResponseTypes.AUCTION_CREATED, ctx.services.auctionService.serializeListing(listing)) },
            onFailure = { error -> HandlerResult.Err(ServerError.fromMessage(error.message ?: "Could not create listing", ErrorCategory.VALIDATION)) }
        )
    }

    private suspend fun bid(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val listingId = payload.str("listingId")
            ?: return HandlerResult.Err(ServerError.fromMessage("listingId required", ErrorCategory.VALIDATION))
        val amount = payload.longOr("amount")
        if (amount <= 0) return HandlerResult.Err(ServerError.fromMessage("amount required", ErrorCategory.VALIDATION))

        val result = ctx.services.auctionService.placeBid(listingId, playerId, amount)
        return result.fold(
            onSuccess = { listing -> HandlerResult.Ok(ResponseTypes.AUCTION_BID_PLACED, ctx.services.auctionService.serializeListing(listing)) },
            onFailure = { error -> HandlerResult.Err(ServerError.fromMessage(error.message ?: "Bid failed", ErrorCategory.VALIDATION)) }
        )
    }

    private suspend fun buyout(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val listingId = payload.str("listingId")
            ?: return HandlerResult.Err(ServerError.fromMessage("listingId required", ErrorCategory.VALIDATION))
        val result = ctx.services.auctionService.buyout(listingId, playerId)
        return result.fold(
            onSuccess = { listing -> HandlerResult.Ok(ResponseTypes.AUCTION_BOUGHT, ctx.services.auctionService.serializeListing(listing)) },
            onFailure = { error -> HandlerResult.Err(ServerError.fromMessage(error.message ?: "Buyout failed", ErrorCategory.VALIDATION)) }
        )
    }

    private suspend fun myListings(ctx: HandlerContext, playerId: String): HandlerResult {
        val listings = ctx.services.auctionService.getPlayerListings(playerId)
        return HandlerResult.Ok(ResponseTypes.AUCTION_MY_LISTINGS, mapOf(
            "listings" to listings.map { ctx.services.auctionService.serializeListing(it) },
            "count" to listings.size
        ))
    }

    private suspend fun cancel(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val listingId = payload.str("listingId")
            ?: return HandlerResult.Err(ServerError.fromMessage("listingId required", ErrorCategory.VALIDATION))
        val result = ctx.services.auctionService.cancelListing(listingId, playerId)
        return result.fold(
            onSuccess = { HandlerResult.Ok(ResponseTypes.AUCTION_CANCELLED, mapOf("listingId" to listingId)) },
            onFailure = { error -> HandlerResult.Err(ServerError.fromMessage(error.message ?: "Cancel failed", ErrorCategory.VALIDATION)) }
        )
    }
}

File: zero-server-csharp/src/ZeroDayMMO.Server/Handlers/AuctionHandler.cs
using System.Text.Json;
using ZeroDayMMO.Server.Services;
using ZeroDayMMO.Shared;

namespace ZeroDayMMO.Server.Handlers;

public class AuctionHandler : IHandler
{
    private readonly AuctionService _auctionService;

    public string MessageType => "auction_list";

    public AuctionHandler(AuctionService auctionService)
    {
        _auctionService = auctionService;
    }

    public async Task<IActionResult> HandleAsync(string connectionId, JsonElement? payload, string? requestId)
    {
        try
        {
            var listings = await _auctionService.SearchListings();
            return new MessageResult("auction_listings", connectionId, new
            {
                listings = listings.Select(l => _auctionService.SerializeListing(l)),
                count = listings.Count
            });
        }
        catch (Exception ex)
        {
            return new MessageResult("error", connectionId, new { message = ex.Message });
        }
    }

    public async Task<IActionResult> HandleSearch(string connectionId, JsonElement? payload, string? requestId)
    {
        try
        {
            var query = payload?.Str("query");
            var rarity = payload?.Str("rarity");
            var slot = payload?.Str("slot");
            var minPrice = payload?.Long("minPrice");
            var maxPrice = payload?.Long("maxPrice");

            var results = await _auctionService.SearchListings(
                query: string.IsNullOrEmpty(query) ? null : query,
                rarity: string.IsNullOrEmpty(rarity) ? null : rarity,
                slot: string.IsNullOrEmpty(slot) ? null : slot,
                minPrice: minPrice.GetValueOrDefault() > 0 ? minPrice : null,
                maxPrice: maxPrice.GetValueOrDefault() > 0 ? maxPrice : null
            );
            return new MessageResult("auction_search_results", connectionId, new
            {
                results = results.Select(l => _auctionService.SerializeListing(l)),
                count = results.Count
            });
        }
        catch (Exception ex)
        {
            return new MessageResult("error", connectionId, new { message = ex.Message });
        }
    }

    public async Task<IActionResult> HandleCreate(string connectionId, JsonElement? payload, string? requestId)
    {
        var playerId = payload?.Str("playerId");
        if (string.IsNullOrEmpty(playerId))
            return new MessageResult("error", connectionId, new { message = "playerId required" });

        var itemName = payload?.Str("itemName");
        var itemId = payload?.Str("itemId");
        var startingBid = payload?.Long("startingBid") ?? 0L;

        if (string.IsNullOrEmpty(itemName))
            return new MessageResult("error", connectionId, new { message = "itemName required" });
        if (string.IsNullOrEmpty(itemId))
            return new MessageResult("error", connectionId, new { message = "itemId required" });
        if (startingBid <= 0)
            return new MessageResult("error", connectionId, new { message = "startingBid required" });

        try
        {
            var listing = await _auctionService.CreateListing(
                sellerId: playerId,
                itemName: itemName,
                itemId: itemId,
                itemRarity: payload?.Str("itemRarity") ?? "Common",
                itemSlot: payload?.Str("itemSlot") ?? "Any",
                itemTier: payload?.Int("itemTier") ?? 1,
                itemLevel: payload?.Int("itemLevel") ?? 1,
                startingBid: startingBid,
                buyoutPrice: payload?.Long("buyoutPrice") ?? 0L,
                durationHours: payload?.Int("durationHours") ?? 24
            );
            return new MessageResult("auction_created", connectionId, _auctionService.SerializeListing(listing));
        }
        catch (Exception ex)
        {
            return new MessageResult("error", connectionId, new { message = ex.Message });
        }
    }

    public async Task<IActionResult> HandleBid(string connectionId, JsonElement? payload, string? requestId)
    {
        var playerId = payload?.Str("playerId");
        if (string.IsNullOrEmpty(playerId))
            return new MessageResult("error", connectionId, new { message = "playerId required" });

        var listingId = payload?.Str("listingId");
        var amount = payload?.Long("amount") ?? 0L;

        if (string.IsNullOrEmpty(listingId))
            return new MessageResult("error", connectionId, new { message = "listingId required" });
        if (amount <= 0)
            return new MessageResult("error", connectionId, new { message = "amount required" });

        try
        {
            var listing = await _auctionService.PlaceBid(listingId, playerId, amount);
            return new MessageResult("auction_bid_placed", connectionId, _auctionService.SerializeListing(listing));
        }
        catch (Exception ex)
        {
            return new MessageResult("error", connectionId, new { message = ex.Message });
        }
    }

    public async Task<IActionResult> HandleBuyout(string connectionId, JsonElement? payload, string? requestId)
    {
        var playerId = payload?.Str("playerId");
        if (string.IsNullOrEmpty(playerId))
            return new MessageResult("error", connectionId, new { message = "playerId required" });

        var listingId = payload?.Str("listingId");
        if (string.IsNullOrEmpty(listingId))
            return new MessageResult("error", connectionId, new { message = "listingId required" });

        try
        {
            var listing = await _auctionService.Buyout(listingId, playerId);
            return new MessageResult("auction_bought", connectionId, _auctionService.SerializeListing(listing));
        }
        catch (Exception ex)
        {
            return new MessageResult("error", connectionId, new { message = ex.Message });
        }
    }

    public async Task<IActionResult> HandleMyListings(string connectionId, JsonElement? payload, string? requestId)
    {
        var playerId = payload?.Str("playerId");
        if (string.IsNullOrEmpty(playerId))
            return new MessageResult("error", connectionId, new { message = "playerId required" });

        try
        {
            var listings = await _auctionService.GetPlayerListings(playerId);
            return new MessageResult("auction_my_listings", connectionId, new
            {
                listings = listings.Select(l => _auctionService.SerializeListing(l)),
                count = listings.Count
            });
        }
        catch (Exception ex)
        {
            return new MessageResult("error", connectionId, new { message = ex.Message });
        }
    }

    public async Task<IActionResult> HandleCancel(string connectionId, JsonElement? payload, string? requestId)
    {
        var playerId = payload?.Str("playerId");
        if (string.IsNullOrEmpty(playerId))
            return new MessageResult("error", connectionId, new { message = "playerId required" });

        var listingId = payload?.Str("listingId");
        if (string.IsNullOrEmpty(listingId))
            return new MessageResult("error", connectionId, new { message = "listingId required" });

        try
        {
            await _auctionService.CancelListing(listingId, playerId);
            return new MessageResult("auction_cancelled", connectionId, new { listingId });
        }
        catch (Exception ex)
        {
            return new MessageResult("error", connectionId, new { message = ex.Message });
        }
    }
}
